File: my-app/src/types/shared.ts
/**
 * Canonical shared types for WarpSense.
 * Import from here. Never redefine these elsewhere.
 * WELD_METRICS must match backend models.shared_enums.WeldMetric values.
 */

// ─── ID Types ───────────────────────────────────────────────────────────────
export type WelderID = string;
export type SessionID = string;
export type SiteID = string;
export type TeamID = string;

// ─── Metric Names ───────────────────────────────────────────────────────────
export type WeldMetric =
  | "angle_consistency"
  | "thermal_symmetry"
  | "amps_stability"
  | "volts_stability"
  | "heat_diss_consistency";

export const WELD_METRICS: WeldMetric[] = [
  "angle_consistency",
  "thermal_symmetry",
  "amps_stability",
  "volts_stability",
  "heat_diss_consistency",
];

export const METRIC_LABELS: Record<WeldMetric, string> = {
  angle_consistency: "Angle Consistency",
  thermal_symmetry: "Thermal Symmetry",
  amps_stability: "Amps Stability",
  volts_stability: "Volts Stability",
  heat_diss_consistency: "Heat Dissipation Consistency",
};

// ─── Severity / Risk ─────────────────────────────────────────────────────────
export type RiskLevel = "ok" | "warning" | "critical";
/** info | warning for session-level; critical reserved in ai-feedback for micro-feedback */
export type FeedbackSeverity = "info" | "warning";

// ─── Metric Score ───────────────────────────────────────────────────────────
export interface MetricScore {
  metric: WeldMetric;
  value: number; // 0–100
  label: string;
}

// ─── Pagination (Batch 2+ list endpoints) ─────────────────────────────────────
export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
}

// ─── Status Enums ───────────────────────────────────────────────────────────
export type AnnotationType =
  | "defect_confirmed"
  | "near_miss"
  | "technique_error"
  | "equipment_issue";

export type CoachingStatus = "active" | "complete" | "overdue";

export type CertificationStatus =
  | "not_started"
  | "on_track"
  | "at_risk"
  | "certified";

File: backend/schemas/shared.py
"""
Pydantic schemas shared across multiple routes/services.
Import from here — never redefine these schemas elsewhere.
"""
from pydantic import BaseModel, field_validator

from models.shared_enums import WeldMetric


class MetricScore(BaseModel):
    metric: WeldMetric
    value: float  # 0.0 – 100.0
    label: str  # Human-readable display name

    @field_validator("value")
    @classmethod
    def value_in_range(cls, v: float) -> float:
        if not 0 <= v <= 100:
            raise ValueError("value must be 0.0–100.0")
        return v


METRIC_LABELS: dict[WeldMetric, str] = {
    WeldMetric.ANGLE_CONSISTENCY: "Angle Consistency",
    WeldMetric.THERMAL_SYMMETRY: "Thermal Symmetry",
    WeldMetric.AMPS_STABILITY: "Amps Stability",
    WeldMetric.VOLTS_STABILITY: "Volts Stability",
    WeldMetric.HEAT_DISS_CONSISTENCY: "Heat Dissipation Consistency",
}


def make_metric_score(metric: WeldMetric, value: float) -> MetricScore:
    return MetricScore(metric=metric, value=value, label=METRIC_LABELS[metric])
